### ISSUE TITLE
Fast sender breaks receiving logic

### ISSUE THREAD

**Issue by kaenganxt** [jhelovuo/RustDDS#241]
I sometimes encounter an issue in my application while receiving data with RustDDS 0.7.11. My application is subscribed to a topic, which is frequently send. I am reading data with `take_next_sample` every 10ms. When two data samples are sent in a 10ms interval, `take_next_sample` doesn't return any data anymore until restarting the program.

### Investigation:
I already investigated the issue, the following are all relevant facts:

The QoS values are as follows:
- RELIABILITY=RELIABLE
- DURABILITY=TRANSIENT_LOCAL
- HISTORY=LAST(1)

The publisher is using FastDDS.

When the problem occurs, the following packets are exchanged (looks good):
```
[Pub->Sub] INFO_DST, INFO_TS, DATA(writerSeqNumber=34), HEARTBEAT(firstAvailableSeqNumber=34,lastSeqNumber=34)
[Sub->Pub] INFO_DST, ACKNACK(Expecting sample 35)
[Pub->Sub] INFO_DST, INFO_TS, DATA(writerSeqNumber=35), HEARTBEAT(firstAvailableSeqNumber=35,lastSeqNumber=35)
[Sub->Pub] INFO_DST, ACKNACK(Expecting sample 36)
```

The problem now seems to be caused by a misunderstanding between the `MessageReceiver` and the `DataReader`.
Steps of the `MessageReceiver`:
- Handle DATA submessage (seq=34)
  - Insert sample into HistoryCache (Had 1 element previously in, now length is 2)
- Handle HEARTBEAT submessage
  - Remove all samples from HistoryCache until `firstAvailableSeqNumber` (so cache length is now 1 again)
- Handle DATA submessage (seq=35)
  - Insert sample into HistoryCache (Length is now 2)
- Handle HEARTBEAT submessage
  - Remove all samples from HistoryCache until `firstAvailableSeqNumber` (so only seq=35 remains in cache)

Afterwards, a new sample is read using `take_next_sample`:

In `DataReader::fill_local_datasample_cache`, the next sample is retrieved from the cache and its sequency number (35) is compared to the previously processed seq=33.
Since reliability is activated and seq=34 is missing, no data is returned.
Since seq=34 was discarded, this will never be received and thus `take_next_sample` will never return a sample again.
See also the TODO here: https://github.com/jhelovuo/RustDDS/blob/692841bfa4040cd53e07054c65b5d0dcc8bffed5/src/dds/with_key/datareader.rs#L718-L720

### Solution?
I am not really sure what the solution to this problem is. I think the possibilities are:

1. Don't remove samples on a HEARTBEAT message. This would partly make sense because the RTPS spec says in 8.3.7.5.4 "Change in state of Receiver": "None", so I don't think deleting entries is the way to go here. On the contrary, this would lead to entries piling up in the cache and they would need to be removed somehow once copied to the datareader's datasample cache.
2. Signal the `DataReader` that certain sequence numbers can be skipped, but I'm not sure if this should be allowed with the RELIABLE QoS.

What do you think on how to proceed? Or did I miss something?

**Comment by jhelovuo** [comment]
Thank you for the report. This one of the best bug reports that I have ever seen, if not the best. Very clearly written, and you have gone through the trouble of analysing what is wrong and even suggesting solutions.

It seems that there is a logic error in processing a received HEARTBEAT. The firstAvailableSeqNumber means that the writer does not have anything before that, but it does not imply that also the reader should discard those samples. The correct solution is along the lines of your suggested Solution 1. And yes, the implication is that HistoryCache needs to be garbage-collected.
 
As you have noticed, the handling of GAP or other cases where a reliable Reader is forced to produce gaps in SequenceNumber sequence, is not complete, and may lead to DataReader locking as you describe. There is a solution in development, but that is targeting RustDDS 0.8, and will introduce other changes as well. However, fixing the HEARTBEAT handling should at least improve the behaviour quite a bit.

 I will see what I can do to the wrong HEARTBEAT processing later this week.

**Comment by jhelovuo** [comment]
This issue seems a bit difficult to fix in the current master / 0.7.x branch.

However, the receive logic has been reworked in the yet-to-be-released [0.8 branch](https://github.com/jhelovuo/RustDDS/tree/0.8), and should work better in this respect. Please try if it works for you.

**Comment by kaenganxt** [comment]
The 0.8 branch works as expected in this case and solves the problem for me. Thank you!
